base/inference.jl

```diff
@@ -1269,7 +1269,6 @@ function tmerge(typea::ANY, typeb::ANY)
         if isa(typea, DataType) && isa(typeb, DataType) && length(typea.parameters) == length(typeb.parameters) && !isvatuple(typea) && !isvatuple(typeb)
             return typejoin(typea, typeb)
         end
-        return Tuple
```

Comment:
Ok I found a good more conservative version of this that doesn't affect performance on #6704.